### DISCUSSION TITLE
Queryable is not object-safe

### DISCUSSION THREAD

**Discussion by jobayle** [blackbeam/rust-mysql-simple#352] [Q&A] [answered]
Dears,

I'd like to make an API that can work with either the Conn or the PooledConn struct, only using functions from the Queryable trait.

Unfortunately Rust does not allow using trait-objects parameters for trait Queryable asit is not Object safe ! (I tried to understand why (I'm fairly new to Rust) I think it's because there is a `where Self: Sized` somewhere, I'm I right?)

What are my option? do I have to create a new type implementing Queryable that can delegate to either Conn or PooledConn?

Thanks!

----

Example:

```rust
pub fn dump_table(conn: &mut Conn, schema: &String, table: &String) -> Result<()> {...} // Allowed
pub fn dump_table(conn: &mut dyn Queryable, schema: &String, table: &String) -> Result<()> {...} // the trait `mysql::prelude::Queryable` cannot be made into an object
pub fn dump_table(conn: Box<dyn Queryable>, schema: &String, table: &String) -> Result<()> {...} // the trait `mysql::prelude::Queryable` cannot be made into an object
```

**Comment by jobayle** [comment]
Just saw that PooledConn is just a wrapper for Conn, so I can only use Conn, sorry for the noise

**Reply by jobayle** [reply]
It seems to make it hard to use a mock object to test my code without having to start a MySQL database though 🙁 
How would you test code working with a Conn?

**Reply by blackbeam** [reply]
Well, I have a database instance on my development machine and I use dockerized mysql in a CI environment 🤷‍♂️

**Reply by jobayle** [reply]
I'll probably do that as well 🙄

**Comment by blackbeam** [comment, marked as answer]
> ```rust
> pub fn dump_table(conn: &mut Conn, schema: &String, table: &String) -> Result<()> {...} // Allowed
> pub fn dump_table(conn: &mut dyn Queryable, schema: &String, table: &String) -> Result<()> {...} // the trait `mysql::prelude::Queryable` cannot be made into an object
> pub fn dump_table(conn: Box<dyn Queryable>, schema: &String, table: &String) -> Result<()> {...} // the trait `mysql::prelude::Queryable` cannot be made into an object
> ```

Never tried dynamic dispatch, actually. Always using generics instead:

```rust
fn by_id<T>(conn: &mut T, id: u64) -> crate::Result<Self>
where
    T: Queryable,
{
    \\ ...
}
```
